### PR TITLE
[TIMOB-25345] Implement extended execution to prevent suspension

### PR DESCRIPTION
**Pull request by infosia**
[TIMOB-25345](https://jira.appcelerator.org/browse/TIMOB-25345)

```js
var win = Ti.UI.createWindow({ backgroundColor: 'gray' });

win.addEventListener('open', function () {
    Ti.UI.Windows.beginExtendedExecution();

    setTimeout(function () {
        Ti.UI.Windows.endExtendedExecution();
    }, 10000);
});

win.open();
```

Note: It seems this is working good on Jenkins server. But from what I observed, desktop app is not suspended even when you lock the screen...so you might not have a good way to test this. Just check if it does not throw exception when you test this.

### CODE REVIEW

File: Source/UI/src/WindowsModule.cpp [hunk 1]
```diff
@@ -127,5 +134,65 @@ namespace TitaniumWindows
...
+#if defined(IS_WINDOWS_10)
+			TITANIUM_EXCEPTION_CATCH_START{
+				endExtendedExecution();
+			} TITANIUM_EXCEPTION_CATCH_END
+#endif
+			return get_context().CreateUndefined();
+		}
+
+#if defined(IS_WINDOWS_10)
+		void WindowsModule::beginExtendedExecution()
```
Comment (sgtcoolguy) [inline Source/UI/src/WindowsModule.cpp]: So, this implementation is working fine for running the unit tests, BUT I would like to see if we can improve the API offered to the developer.

First, this should likely hang off Ti.App.Windows, not Ti.UI.Windows.

Second, I would think we'd want to give some mechanism to return the result of trying to request the extension, and to register a callback for if/when the extension is revoked. Additionally, right now this assumes the reason is "unspecified". We should probably expose the possible reasons.

Right now you simply request the extension and have no return value to indicate if it worked. You also have no way to handle when it gets revoked.

Can we take an optional object parameter that could contain the reason, the callback for the result, and the callback for revocation?

i.e.
```javascript
Ti.App.Windows.requestExtendedExecution({
  reason: Ti.App.Windows.EXTENDED_EXECUTION_REASON_SAVING_DATA,
  result: function (granted) { // granted is a boolean value, true or false
  },
  revoked: function(reason) { // reason is an enumerated constant from https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.extendedexecution.extendedexecutionrevokedreason
  }
});
```

Or maybe just avoid the object and do:
```javascript
Ti.App.Windows.requestExtendedExecution(Ti.App.Windows.EXTENDED_EXECUTION_REASON_SAVING_DATA, function (granted) { // granted is a boolean value, true or false
  }, function(reason) { // reason is an enumerated constant from https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.extendedexecution.extendedexecutionrevokedreason
  }
});
```
With the first arg optional, defaults to UNSPECIFIED, second arg is optional, callback for result, third is optional callback for revocation?

**Comment by infosia** [comment]
@sgtcoolguy Updated, implemented like below.

```js
var win = Ti.UI.createWindow({ backgroundColor: 'green' });

win.addEventListener('open', function () {
    Ti.App.Windows.requestExtendedExecution(
        Ti.App.Windows.EXTENDED_EXECUTION_REASON_UNSPECIFIED,
        function (granted) {
            alert('granted = ' + granted);
            Ti.App.Windows.closeExtendedExecution();
        },
        function (reason) {
            alert('revoked reason = ' + reason);
        }
    );
});

win.open();
```

File: Source/Ti/src/App/WindowsModule.cpp [hunk 2]
```diff
@@ -43,12 +44,140 @@ namespace TitaniumWindows
...
+			return get_context().CreateNumber(1);
+#endif
+		}
+
+		TITANIUM_FUNCTION(WindowsModule, requestExtendedExecution)
+		{
+#if defined(IS_WINDOWS_10)
+			using namespace Windows::ApplicationModel::ExtendedExecution;
+
+			ENSURE_OPTIONAL_INT_AT_INDEX(reason, 0, 0);
```
Comment (sgtcoolguy) [inline Source/Ti/src/App/WindowsModule.cpp]: So I assume that 0 == Unspecified, but it leaves me a little concerned that we're not explicitly using the value of `Windows::ApplicationModel::ExtendedExecution::ExtendedExecutionReason::Unspecified` here.

Comment (infosia) [reply Source/Ti/src/App/WindowsModule.cpp]: Makes sense, updated 👍

**Comment by infosia** [comment]
@ewanharris I think that makes sense. Making the parameter `Object` should make the developer's option more flexible and more unambiguous.

Fixed, now it's like below. Every params for `requestExtendedExecution` is optional now.

```js
var win = Ti.UI.createWindow({ backgroundColor: 'green' });
var label = Ti.UI.createLabel();

var counter = 0;
var handle;
var swtch = Ti.UI.createSwitch({
    top: 100,
    title: 'Update in background?'
});

swtch.addEventListener('change', function (e) {
    if (e.value) {
        Ti.App.Windows.requestExtendedExecution({
            reason: Ti.App.Windows.EXTENDED_EXECUTION_REASON_UNSPECIFIED,
            result: function (granted) {
                Ti.API.info('granted = ' + granted);
                if (granted) {
                    handle = setInterval(function () {
                        counter++;
                        label.text = counter
                    }, 100);
                } else {
                    alert('Couldnt allow extended execution')
                }
            },
            revoked: function (reason) {
                Ti.API.info('revoked')
            }
        });
    } else {
        Ti.API.info('Stopping execution');
        Ti.App.Windows.closeExtendedExecution();
        clearInterval(handle);
    }
});

Ti.App.addEventListener('resume', function () {
    console.log('resumed');
})

win.add(label);
win.add(swtch);
win.open();
```

**Pull request merged**